refactor(transformer): `HelperLoader`: add import immediately

overlookmotel opened the pull request:
Helper loader does not need to store details for import statements and then add them in an `exit_program` visitor. It can add them immediately.

That means `HelperLoader` transform can be removed, and its public methods are implemented directly on `TransformCtx`.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6601?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6601** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6601?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#6599** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6599?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6596** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6596?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6595** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6595?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
I think this is *probably* an improvement, but not 100% sure. Please let me know if you think it was better before.

If we do merge this, then I'll do a follow-up to move the implementation out of the "common" transforms directory and into `context`, since it's no longer a transform.

Dunqing commented on crates/oxc_transformer/src/common/helper_loader.rs:
```diff
@@ -182,16 +164,19 @@ impl<'a> HelperLoaderStore<'a> {
     }
+}
 
+// Public methods implemented directly on `TransformCtx`, as they need access to `TransformCtx::module_imports`.
+impl<'a> TransformCtx<'a> {
```
Do you think this is a good place to support helper-related methods for `TransformCtx`

Dunqing replied:
Just saw your [comment](https://github.com/oxc-project/oxc/pull/6601#issuecomment-2414554481).

Dunqing commented:
> I think this is _probably_ an improvement, but not 100% sure. Please let me know if you think it was better before.
> 
> If we do merge this, then I'll do a follow-up to move the implementation out of the "common" transforms directory and into `context`, since it's no longer a transform.

Yes, this is indeed a performance improvement. I am just a little confused, here added some methods for `TransformerCtx`.  We can merge this. I am looking forward to your follow-up PRs

overlookmotel commented:
### Merge activity

* **Oct 16, 9:01 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
refactor(transformer): `HelperLoader`: add import immediately (#6601)

Helper loader does not need to store details for import statements and then add them in an `exit_program` visitor. It can add them immediately.

That means `HelperLoader` transform can be removed, and its public methods are implemented directly on `TransformCtx`.

overlookmotel commented:
> I am just a little confused, here added some methods for `TransformerCtx`.

The idea is:

Previously, `HelperLoaderStore` stored the details of imports internally, and then at the very end of traversal, in `exit_program`, it called `ModuleImportsStore` to add those imports.

I realized that there was no need to wait until the end to pass the details of imports to `ModuleImportsStore`. It could just do it immediately when you request a helper.

But to do that, `load` needs access *both* `HelperLoaderStore` and `ModuleImportsStore`, which are both properties of `TransformCtx`.

So I moved `load` method to be a method of `TransformCtx`, so it can access both.

I don't know if that makes any more sense...
